src/nnet3/online-nnet3-decodable-simple.cc

```diff
@@ -0,0 +1,210 @@
+  int32 features_ready = features_->NumFramesReady();
+  if (features_ready == 0)
+    return 0;
+  bool input_finished = features_->IsLastFrame(features_ready - 1);
+  if (opts_.pad_input) {
+    // normal case... we'll pad with duplicates of first + last frame to get the
+    // required left and right context.
+    if (input_finished) return NumSubsampledFrames(features_ready);
+    else return std::max<int32>(0, NumSubsampledFrames(features_ready - right_context_));
+  } else {
+    return std::max<int32>(0, NumSubsampledFrames(features_ready - right_context_ - left_context_));
+  }
+}
+
+int32 DecodableNnet3SimpleOnline::NumSubsampledFrames(int32 num_frames) const {
+  return (num_frames) / opts_.frame_subsampling_factor;
+}
+
+void DecodableNnet3SimpleOnline::ComputeForFrame(int32 subsampled_frame) {
+  int32 features_ready = features_->NumFramesReady();
+  bool input_finished = features_->IsLastFrame(features_ready - 1);  
+  KALDI_ASSERT(subsampled_frame >= 0);
+  if (subsampled_frame >= begin_frame_ &&
+      subsampled_frame < begin_frame_ + scaled_loglikes_.NumRows())
+    return;
+  KALDI_ASSERT(subsampled_frame < NumFramesReady());
+
+  int32 subsample = opts_.frame_subsampling_factor;
+
+  int32 input_frame_begin;
+  if (opts_.pad_input)
+    input_frame_begin = subsampled_frame * subsample  - left_context_;
+  else
+    input_frame_begin = subsampled_frame * subsample;
+  int32 max_possible_input_frame_end = features_ready /* - ( features_ready - right_context_) % subsample */;
+  if (input_finished && opts_.pad_input)
+    max_possible_input_frame_end += right_context_;
+  int32 input_frame_end = std::min<int32>(max_possible_input_frame_end,
+                                          input_frame_begin +
+                                          left_context_ + right_context_ +
+                                          opts_.max_nnet_batch_size);
+  KALDI_ASSERT(input_frame_end > input_frame_begin);
+  Matrix<BaseFloat> features(input_frame_end - input_frame_begin,
+                             feat_dim_);
+  for (int32 t = input_frame_begin; t < input_frame_end; t++) {
+    SubVector<BaseFloat> row(features, t - input_frame_begin);
+    int32 t_modified = t;
+    // The next two if-statements take care of "pad_input"
+    if (t_modified < 0)
+      t_modified = 0;
+    if (t_modified >= features_ready)
+      t_modified = features_ready - 1;
+    features_->GetFrame(t_modified, &row);
+  }
+
+  int32 num_subsampled_frames = NumSubsampledFrames(input_frame_end - input_frame_begin -
+          left_context_ - right_context_);
+  // I'm not checking if the input feature vector is ok.
+  // It should be done, but I'm not sure if it is the best place. 
+  // Maybe a new "nnet3 feature pipeline"?
+  int32 mfcc_dim = am_nnet_.GetNnet().InputDim("input");
+  int32 ivector_dim = am_nnet_.GetNnet().InputDim("ivector");
+  // MFCCs in the left chunk
+  SubMatrix<BaseFloat> mfcc_mat = features.ColRange(0,mfcc_dim);
```

Comment:
this should not use the = syntax but the constructor () syntax because if it were assignment, the code would not work.
 (probably C++ automatically makes it construction but let's be explicit).